stan/math/rev/fun/dot_product.hpp

```diff
@@ -229,11 +30,32 @@ inline return_type_t<T1, T2> dot_product(const T1* v1, const T2* v2,
  * @return Dot product of the vectors.
  * @throw std::domain_error if sizes of v1 and v2 do not match.
  */
-template <typename T1, typename T2, typename = require_any_var_t<T1, T2>>
-inline return_type_t<T1, T2> dot_product(const std::vector<T1>& v1,
-                                         const std::vector<T2>& v2) {
+template <typename T1, typename T2, require_all_container_t<T1, T2>* = nullptr,
+          require_any_vt_var<T1, T2>* = nullptr>
+inline return_type_t<T1, T2> dot_product(const T1& v1, const T2& v2) {
   check_matching_sizes("dot_product", "v1", v1, "v2", v2);
```

Comment:
[optional]

I think this size check is always extraneous because the prim version always checks.

But this check is cheap and would avoid allocating more memory before throwing.

I guess keep it?

Reply:
It does not matter much. I am fine with keeping it.